### PR TITLE
refactor: update style guide and rewrite output formatting page

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind user-interface

> /kind content

> /kind translation

> /kind event

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area blog

/area documentation

> /area community

**What this PR does / why we need it**:

refactor: update style guide and rewrite output formatting page
CC @Andreagit97 and @LucaGuerra @leogr thanks!

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by leogr** [comment]
cc @LucaGuerra @mikegcoleman

File: content/en/docs/outputs/formatting.md [hunk 1]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+
+Do you have an even more customized use case? We have you covered! We also provide a `-p` flag where you can define custom additional output fields to be included in each rule.
+
+## Example Rule
 
+```yaml
+- rule: Drop and execute new binary in container
+  desc: >
+    SKIPPED
+  condition: >
```
Comment (leogr) [inline content/en/docs/outputs/formatting.md:38]:
```suggestion
- rule: Drop and execute new binary in container
  desc:
    SKIPPED
  condition:
```

`>` would introduce unnecessary `\n` and should be discouraged in examples.

File: content/en/docs/outputs/formatting.md [hunk 2]
```diff
@@ -7,48 +7,96 @@ aliases:
...
 
-## Examples
+### Scenario 1
 
-Here are some examples of Falco command lines, output strings in rules, and the resulting output:
+The rule outputs include `%container.info`, no command-line flags:
 
-### Output contains `%container.info`
+```bash 
+sudo /usr/bin/falco  -c /etc/falco/falco.yaml -r falco_rules_test.yaml -p
```
Comment (leogr) [inline content/en/docs/outputs/formatting.md:53]:
```suggestion
sudo /usr/bin/falco  -c /etc/falco/falco.yaml -r falco_rules_test.yaml
```
?

File: content/en/docs/rules/style-guide.md [hunk 3]
```diff
@@ -227,7 +235,7 @@ Some rules are more specific signatures, while others focus on behavior-based de
 
 ### Rules Loading
 
-Refer to the up-to-date description in the [falco.yaml](https://github.com/falcosecurity/falco/blob/master/falco.yaml) file for `rules_file` to understand in which order rules are loaded. Keep in mind that Falco applies rules per event type on a "first match wins" basis.
+Refer to the up-to-date description in the [falco.yaml](https://github.com/falcosecurity/falco/blob/master/falco.yaml) file for `rules_file` to understand in which order rules are loaded. Keep in mind that Falco by default applies rules per event type on a "first match wins" basis. Starting from Falco Release 0.36.0, you have the option to modify the configuration to `rule_matching: all`. This change ensures that rules sharing the same event type cannot override each other, preventing inconsistent logging. Be aware, though, that this modification may lead to increased CPU usage.
```
Comment (leogr) [inline content/en/docs/rules/style-guide.md:238]: Maybe we can explain the effect of enabling `rule_matching: all` or pointing to the relevant documentation

**Review by leogr — COMMENTED** [review]
Overall, SGMT!

Just a couple of suggestions

File: content/en/docs/outputs/formatting.md [hunk 4]
```diff
@@ -7,48 +7,96 @@ aliases:
...
-* If the format string does not contain `%container.info`, and one of `-pk`/`-pc` was provided, that is added to the end of the formatting string.
+# Remove output fields
+output: %proc.sname %user.name
 
-* If `-p` was specified with a general value (i.e. not `-pk`/`-pc`), the value is simply added to the end and any `%container.info` is replaced with the generic value.
+# Add more output fields again
+output: %proc.sname %user.name %proc.is_exe_upper_layer
+```
+
+However, if you want to add the same output field to every rule, doing so manually is tedious. You might seek a shortcut to generically add specific output fields to every rule. The following explanation details how to achieve this.
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:26]:
```suggestion
However, adding the same output field to multiple rules manually can be tedious. Thankfully Falco has a shortcut to make this eaier.

File: content/en/docs/outputs/formatting.md [hunk 5]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+
+Falco inherently supports event decoration with associated [Container](https://falco.org/docs/reference/rules/supported-fields/#field-class-container) and [Kubernetes Fields](https://falco.org/docs/reference/rules/supported-fields/#field-class-k8s). To accomplish this, you'll need a special placeholder field in your rules' output fields, denoted as `%container.info`, and you'll need to run Falco with either the `-pk` or `-pc` command-line option. If your rule does not include `%container.info` no container or Kubernetes fields will be added.
+
+Do you have an even more customized use case? We have you covered! We also provide a `-p` flag where you can define custom additional output fields to be included in each rule.
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:30]:
```suggestion
If you have an even more customized use case, Falco also provides a `-p` flag where you can define additional custom output fields to be included in each rule.
```

File: content/en/docs/outputs/formatting.md [hunk 6]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+  output: Executing binary not part of base image (proc_sname=%proc.sname user=%user.name process=%proc.name proc_exepath=%proc.exepath parent=%proc.pname command=%proc.cmdline terminal=%proc.tty %container.info)
+  priority: CRITICAL
+  tags: [maturity_stable, container, process, mitre_persistence, TA0003, PCI_DSS_11.5.1]
+```
 
-## Examples
+### Scenario 1
 
-Here are some examples of Falco command lines, output strings in rules, and the resulting output:
+The rule outputs include `%container.info`, no command-line flags:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:50]: I don't know what we're saying here with the ", no command-line" flag - do we mean falco wasn't started with -pk or -pc?

If so, maybe it sounds better to say "The rule outputs include `%container.info`,  but Falco was started without the -pk or -pc command-line flags"

File: content/en/docs/outputs/formatting.md [hunk 7]
```diff
@@ -7,48 +7,96 @@ aliases:
...
 
-### Output contains `%container.info`
+```bash 
+sudo /usr/bin/falco  -c /etc/falco/falco.yaml -r falco_rules_test.yaml -p
 ```
-output: "Namespace change (setns) by unexpected program (user=%user.name command=%proc.cmdline parent=%proc.pname %container.info)"
 
-$ falco
-15:42:35.347416068: Warning Namespace change (setns) by unexpected program (user=root command=test_program parent=hyperkube k8s-kubelet (id=4a4021c50439))
+Nevertheless, we continue to output `%container.id` and `%container.name`:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:56]:
```suggestion
In this case Falco will output `%container.id` and `%container.name` but no other container metatdat will be displayed: 
```

File: content/en/docs/outputs/formatting.md [hunk 8]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+03:00:45.104332605: Critical Executing binary not part of base image (proc_sname=bash user=root process=sleep proc_exepath=/tmp/sleep parent=bash command=sleep 10000 terminal=34816 container_id=0fdb3cd5b5fc container_name=optimistic_newton)
+```
 
-$ falco -pk -k <k8s api server url>
-15:42:35.347416068: Warning Namespace change (setns) by unexpected program (user=root command=test_program parent=hyperkube k8s.pod=jclient-3160134038-qqaaz container=4a4021c50439)
+### Scenario 2
 
-$ falco -p "This is Some Extra" -k <k8s api server url>
-15:42:35.347416068: Warning Namespace change (setns) by unexpected program (user=root command=test_program parent=hyperkube k8s-kubelet (id=4a4021c50439)) This is Some Extra
+The rule outputs include `%container.info`, and to enable this functionality, you run Falco with the `-pc` flag:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:64]:
```suggestion
The rule outputs include `%container.info`, and Falco is started with the `-pc` flag: 
```

File: content/en/docs/outputs/formatting.md [hunk 9]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+
+```bash 
+sudo /usr/bin/falco  -c /etc/falco/falco.yaml -r falco_rules_test.yaml -pc
 ```
 
-### Output does not contain `%container.info`
+Output includes the default container fields:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:70]:
```suggestion
The output includes the default container fields:
```

File: content/en/docs/outputs/formatting.md [hunk 10]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+```bash
+03:02:52.019002207: Critical Executing binary not part of base image (proc_sname=bash user=root process=sleep proc_exepath=/tmp/sleep parent=bash command=sleep 10000 terminal=34816 container_id=0fdb3cd5b5fc container_image=ubuntu container_image_tag=latest container_name=optimistic_newton)
 ```
-output: "File below a known binary directory opened for writing (user=%user.name command=%proc.cmdline file=%fd.name)"
 
-$ falco
-15:50:18.866559081: Warning File below a known binary directory opened for writing (user=root command=touch /bin/hack file=/bin/hack) k8s-kubelet (id=4a4021c50439)
+### Scenario 3
+
+The rule outputs include `%container.info`, and to enable this functionality, you run Falco with the `-pk` flag:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:78]:
```suggestion
The rule outputs include `%container.info`, and Falco is started with the `-pk` flag:
```

File: content/en/docs/outputs/formatting.md (hunk 6 above)
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:50]:
```suggestion
The rule outputs include `%container.info`, but Falco is started without any command line flags:
```

File: content/en/docs/outputs/formatting.md [hunk 11]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+
+```bash 
+sudo /usr/bin/falco  -c /etc/falco/falco.yaml -r falco_rules_test.yaml -pk
+```
+
+Output includes the default container plus Kubernetes fields:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:84]:
```suggestion
Output includes the default container fields and the default Kubernetes fields:
```

File: content/en/docs/outputs/formatting.md [hunk 12]
```diff
@@ -7,48 +7,96 @@ aliases:
...
+
+```bash
+03:03:23.573329751: Critical Executing binary not part of base image (proc_sname=bash user=root process=sleep proc_exepath=/tmp/sleep parent=bash command=sleep 10000 terminal=34816 container_id=0fdb3cd5b5fc container_image=ubuntu container_image_tag=latest container_name=optimistic_newton k8s_ns=my_ns k8s_pod_name=my_pod_name)
+```
+
+### Scenario 4
+
+The rule outputs include `%container.info`, and you run Falco with the `-p` flag with providing custom output fields:
```
Comment (mikegcoleman) [inline content/en/docs/outputs/formatting.md:92]:
```suggestion
The rule outputs include `%container.info`, and you run Falco with the `-p` flag while providing custom output fields:
```

**Review by mikegcoleman — COMMENTED** [review]
A few comments, feel free to ignore as they are just around phrasing and grammar.

File: content/en/docs/rules/style-guide.md [hunk 13]
```diff
@@ -65,7 +65,7 @@ evt.type!=execve
 ```
 
 - To maintain performance, avoid mixing unrelated event types in one rule. Typically, only variants should be mixed together, for example: `evt.type in (open, openat, openat2)`.
-- The best practice and requirement for upstream rules are to only define positive `evt.type` expressions. Using `evt.type!=open`, for example, would imply each of the [supported syscalls](https://falcosecurity.github.io/libs/report/), resulting in a significant performance penalty. For more information, read the [Adaptive Syscalls Selection in Falco](/blog/adaptive-syscalls-selection/) blog post.
+- The best practice and requirement for upstream rules are to only define positive `evt.type` expressions. Using `evt.type!=open`, for example, would imply each of the [Supported Syscalls](https://falcosecurity.github.io/libs/report/), resulting in a significant performance penalty. For more information, read the [Adaptive Syscalls Selection in Falco](/blog/adaptive-syscalls-selection/) blog post.
```
Comment (mikegcoleman) [inline content/en/docs/rules/style-guide.md:68]: I'm not sure imply is the right word. Would "activate" be better?

File: content/en/docs/rules/style-guide.md [hunk 14]
```diff
@@ -65,7 +65,7 @@ evt.type!=execve
...
 
 
 - After the `evt.type` filter, place your mainly positive filters to efficiently eliminate the most events step by step. An exception to this rule is the `container` macro, which can quickly eliminate many events. Therefore, the guiding principle of "divide and conquer" commonly used in database query recommendations, also applies to Falco's filter statements.
```
Comment (mikegcoleman) [inline content/en/docs/rules/style-guide.md:71]: I'm not sure what "mainly positive" means.

File: content/en/docs/rules/style-guide.md [hunk 15]
```diff
@@ -138,16 +138,24 @@ High-volume syscalls can increase CPU usage and cause kernel side event drops in
...
+For each rule include the critical fields listed below related to process and user information, as well as the actual event type. For example, the `tty` field (terminal) can help determine if the process ran in an interactive shell. Additionally, examining the exepath alongside the process name provides insights into whether the binary might be located in more suspicious folders like `tmp`. Understanding the direct parent process is vital for basic process lineage analysis.
 
 ```yaml
-evt_type=%evt.type user=%user.name user_uid=%user.uid user_loginuid=%user.loginuid process=%proc.name proc_exepath=%proc.exepath parent=%proc.pname command=%proc.cmdline terminal=%proc.tty exe_flags=%evt.arg.flags
+evt_type=%evt.type user=%user.name user_uid=%user.uid user_loginuid=%user.loginuid process=%proc.name proc_exepath=%proc.exepath parent=%proc.pname command=%proc.cmdline terminal=%proc.tty
 ```
 
+For rules involving only `spawned_process`, please also include `%evt.arg.flags` in the output fields. If a rule involves multiple syscalls beyond `spawned_process`, do not include `%evt.arg.flags`. Our agreement specifies excluding output fields that may be empty or `<NA>` in certain cases.
```
Comment (mikegcoleman) [inline content/en/docs/rules/style-guide.md:151]: "Our agreement" - what agreement? I'm not sure what this means?

File: content/en/docs/rules/style-guide.md [hunk 16]
```diff
@@ -167,15 +175,15 @@ When writing a rule for container workloads, you should include the fields we au
 %container.info
 ```
 
-This special placeholder will be automatically replaced with common container-related fields and, when used in conjunction with `-pk` (see the [output formatting](/docs/outputs/formatting/) page for more details) it will also add [basic Kubernetes fields](/docs/reference/rules/supported-fields/#field-class-k8s) (this does *not* require the *Kubernetes Metadata Enrichment*, i.e. `-k/-K`, functionality to be enabled). Using `%container.info` in conjunction with `-pk` is equivalent to:
+This special placeholder will be automatically replaced with common container-related fields and, when used in conjunction with `-pk` (see the [Output Formatting](/docs/outputs/formatting/) page for more details) it will also add [Basic Kubernetes Fields](/docs/reference/rules/supported-fields/#field-class-k8s). Using `%container.info` in conjunction with `-pk` is equivalent to:
```
Comment (mikegcoleman) [inline content/en/docs/rules/style-guide.md:178]: Does it work with both -pk and -pc? Or is that not relevant to this section?

**Review by mikegcoleman — COMMENTED** [review]
Just a few points I wasn't sure on, and might need clarifying/

File: content/en/docs/outputs/formatting.md (hunk 2 above)
Comment (incertum) [reply content/en/docs/outputs/formatting.md:53]: jeez Melissa ...

File: content/en/docs/rules/style-guide.md (hunk 16 above)
Comment (incertum) [reply content/en/docs/rules/style-guide.md:178]: @leogr more ideas here?

**Comment by incertum** [comment]
Thank you both for the great suggestion, perhaps take another look now?

**Comment by mikegcoleman** [comment]
Based on our timeline - I am fine pushing these changes as they are, and we can clean up the language post-launch. 

LGTM

File: content/en/docs/rules/style-guide.md (hunk 16 above)
Comment (leogr) [reply content/en/docs/rules/style-guide.md:178]: Yes, it works both for `-pk` and `-pc`. Not sure if it is really relevant here.

To summarize:

`-pc` is `container_id=%container.id container_image=%container.image.repository container_image_tag=%container.image.tag container_name=%container.name`

`-pk` is `-pc` plus `k8s_ns=%k8s.ns.name k8s_pod_name=%k8s.pod.name`

(implementation is here: https://github.com/falcosecurity/falco/blob/master/userspace/falco/app/actions/init_falco_engine.cpp#L24)

In both cases:
> If a rule's output contains %container.info, it will be replaced with the corresponding details. Otherwise, these details will be directly appended to the rule's output.

This is documented in the [Output Formatting](https://falco.org/docs/outputs/formatting/) page (and in the Falco inline command line help), so I'm not sure if it's worth duplicateing the content here.

Comment (incertum) [reply content/en/docs/rules/style-guide.md]: Tried to simplify it let me know

**Review by leogr — APPROVED** [review]
LGTM now.

@incertum this seems not tied to 0.37 (unless I miss some detail), so we may merge it now. Correct?
/hold just for double-checking :point_up:

**Comment by incertum** [comment]
Yes good to merge now!
